feat(html,asset): improve asset resolution

grievouz opened the pull request:
<!-- Thank you for contributing! -->

### Description

* Expand on the default set of tags to convert into imports:
Closes https://github.com/vitejs/vite/issues/7362
Closes https://github.com/vitejs/vite/issues/5098

* Use Rollup to resolve import urls (_edit by @patak-dev: To be done in a separate PR_) 
https://github.com/vitejs/vite/issues/3000

<!-- ### Additional context  -->

<!-- e.g. is there anything you'd like reviewers to focus on? -->

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [x] New Feature
- [x] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

grievouz commented:
> * [ ]  Ideally, include relevant tests that fail without this PR but pass with it.

I didn't look further into it but if needed/wanted I can implement tests to cover the tagged issues.

poyoho commented on packages/vite/src/node/plugins/html.ts:
```diff
@@ -116,14 +116,261 @@ export function addToHTMLProxyTransformResult(
+        relProp.value?.content &&
+        [
+          'stylesheet',
+          'icon',
+          'shortcut icon',
+          'mask-icon',
+          'apple-touch-icon',
+          'apple-touch-icon-precomposed',
+          'apple-touch-startup-image',
+          'manifest',
+          'prefetch',
+          'preload'
+        ].includes(relProp.value.content)
+      ) {
+        return true
+      }
+
+      const itempropProp = findProp(node, 'itemprop')
+      if (
+        itempropProp &&
+        itempropProp.type === NodeTypes.ATTRIBUTE &&
+        itempropProp.value?.content &&
+        [
+          'image',
+          'logo',
+          'screenshot',
+          'thumbnailurl',
+          'contenturl',
+          'downloadurl',
+          'duringmedia',
+          'embedurl',
+          'installurl',
+          'layoutimage '
+        ].includes(itempropProp.value.content)
+      ) {
+        return true
+      }
+
+      return false
+    }
```
I think this can also be configured here. I don't know why it should be turned into a method. I would appreciate it if you could tell me

```js
{
  rel: ['stylesheet','icon','shortcut icon','mask-icon','apple-touch-icon','apple-touch-icon-precomposed','apple-touch-startup-image','manifest','prefetch','preload'],
  itemprop: ['image','logo','screenshot','thumbnailurl','contenturl','downloadurl','duringmedia','embedurl','installurl','layoutimage']
}
```

grievouz replied:
I didn't give much thought to how it should be configured. I basically replicated the API from [html-loader](https://webpack.js.org/loaders/html-loader/), as proposed by https://github.com/vitejs/vite/issues/7362#issuecomment-1071989522. Furthermore, the approach is most likely a matter of personal preference. If the array-based configuration fits better into vites config schema then that's how it should be done, so as not to create unnecessary complexity where it isn't required.

poyoho replied:
Since the processing I've seen is all the same, I'm guessing it can be done this way. 🫣  This reduces a lot of code.

patak-dev replied:
I would also prefer we simplify the code. But the functional form would allow for a more flexible configuration point since the PR is exposing `build.htmlAssetSources`. I'm not sure we would like to expose this complexity it is better to leave it to userland. Maybe we could only correct the list as a first step. Added the issue for discussion in a future team meeting.

bluwy replied:
From my experience building [svelte-preprocess-import-assets](https://github.com/bluwy/svelte-preprocess-import-assets), this is actually the only way to ensure correctness (with a filter). I tried simplifying it but it isn't quite achievable if we follow the html spec.

bluwy replied:
> since the PR is exposing `build.htmlAssetSources`

It doesn't look like the PR expose it now (maybe I've missed a discussion) 🤔

patak-dev replied:
Ya, the change to remove it was force pushed. We can still discuss it in the meeting, the PR was exposing `build.htmlAssetSources` and letting you define whatever you want instead of the default `defaultHtmlAssetSources` array. The API was exactly the one of that array.

grievouz replied:
Sorry, my mistake. I discovered some old incorrectly merged changes while rebasing the branch earlier. However, I have the changes that expose `build.htmlAssetSources` saved locally. I'll commit them later so that the discussion has a point of reference.

poyoho commented on packages/vite/src/node/plugins/html.ts:
```diff
@@ -116,14 +116,261 @@ export function addToHTMLProxyTransformResult(
+          'duringmedia',
+          'embedurl',
+          'installurl',
+          'layoutimage '
```
a typo 'layoutimage'

grievouz replied:
Gosh My apologies, I think I corrected that one around ten times while messing around with git.

poyoho commented on packages/vite/src/node/plugins/html.ts:
```diff
@@ -317,33 +564,28 @@ export function buildHtmlPlugin(config: ResolvedConfig): Plugin {
+            // assetsUrl may be encodeURI
+            const url = decodeURI(assetSourceAttr.value!.content)
+            if (!isExcludedUrl(url)) {
+              if (node.tag === 'link' && isCSSRequest(url)) {
```
Should prefetch / preload be loaded directly?

patak-dev replied:
Is this question unrelated to the PR, no? Could you explain a bit more?

patak-dev commented on packages/vite/src/node/plugins/asset.ts:
```diff
@@ -347,6 +347,10 @@ export async function urlToBuiltUrl(
   config: ResolvedConfig,
   pluginContext: PluginContext
 ): Promise<string> {
+  const resolved = await pluginContext.resolve(url, importer, {})
+  if (resolved) {
+    url = path.relative(path.dirname(importer), resolved.id)
+  }
```
Let's remove this change so we can move forward with the improved htmlAssetSources. This change may be more involved, better to review it in its own PR.

grievouz replied:
I agree; I had to tweak a _couple_ of things to get the async part to function in some parts of the codebase.

grievouz committed:
feat(html): improve asset loading

Add logic for selecting source attributes and extend build options to allow for configuration.

bluwy commented on packages/vite/src/node/plugins/html.ts:
```diff
@@ -121,14 +121,261 @@ export function addToHTMLProxyTransformResult(
+  }
+}
+
+export const defaultHtmlAssetSources: HtmlAssetSource[] = [
```
I think it would be great to extract this and it's related types/functions as it's own file as this is HUGE. Maybe we can put it in the `src/node` folder

grievouz committed:
feat(html): add htmlAssetSources to BuildOptions

floriangosse commented:
Is there any update on this? Do you need some support for this?

benmccann commented:
There's a merge conflict for this PR that would need to be resolved

aleclarson commented:
This looks great. Are you still willing to carry this forward, @grievouz? We would love to ship this in Vite 4.

bluwy commented:
Looks like we had discussed this in a couple meetings ago and forgot to give an update. We think this feature is great for Vite 4, but should be shipped without an option first.

bluwy commented:
Closing this one for now in favour of https://github.com/vitejs/vite/pull/11138
